File: web/src/lib/plan.ts
import type { TPlan, TSubscriptionStatus, TUser } from '@/types/models';

const ACTIVE_STATUSES: TSubscriptionStatus[] = ['active', 'trialing'];

export const getActivePlan = (user: TUser | null | undefined): TPlan => {
    if (!user || !isPlanActive(user)) {
        return 'free';
    }
    return user.plan ?? 'free';
};

export const isPlanActive = (user: TUser | null | undefined): boolean => {
    if (!user) {
        return false;
    }
    const status = user.subscription_status;
    if (!status || !ACTIVE_STATUSES.includes(status)) {
        return false;
    }
    if (user.subscription_cancel_at && new Date(user.subscription_cancel_at) <= new Date()) {
        return false;
    }
    return true;
};

File: internal/billing/plan.go
package billing

import (
	"time"

	"github.com/pocketbase/pocketbase/core"
)

func GetActivePlan(user *core.Record) string {
	if user == nil || !IsPlanActive(user) {
		return "free"
	}
	plan := user.GetString("plan")
	if plan == "" {
		return "free"
	}
	return plan
}

func IsPlanActive(user *core.Record) bool {
	if user == nil {
		return false
	}
	var activeStatuses = map[string]bool{
		"active":   true,
		"trialing": true,
	}
	if !activeStatuses[user.GetString("subscription_status")] {
		return false
	}
	cancelAt := user.GetString("subscription_cancel_at")
	if cancelAt != "" {
		t, err := time.Parse(time.RFC3339, cancelAt)
		if err == nil && !t.After(time.Now().UTC()) {
			return false
		}
	}
	return true
}
